src/types.ts

```diff
@@ -82,8 +86,16 @@ export interface SearchRequest {
 
 export type Hit<T> = T & { _formatted?: T }
 
-export interface SearchResponse<T = any> {
-  hits: Array<Hit<T>>
+export interface SearchResponse<T, P extends SearchParams<T>> {
+  hits: P['attributesToRetrieve'] extends keyof T
+    ? Array<
+        Hit<
+          Pick<T, Exclude<keyof T, Exclude<keyof T, P['attributesToRetrieve']>>>
+        >
+      >
+    : P['attributesToRetrieve'] extends Array<infer K>
+    ? Array<Hit<Pick<T, Exclude<keyof T, Exclude<keyof T, K>>>>>
+    : Array<Hit<T>>
```

Comment:
What happens if `attributesToRetrieve` has a key that is not present in any of the returned documents.

Exemple: 
```javascript
const document = {
 id: string, 
 title: string,
 description: string
}

index.search<document>('batman', {
 attributesToRetrieve: ["description", "id", "title"]
}
```

The only document returned in `hits` has no `description` field.
```
{
id: 1,
title: "batman"
}
```

How will this impact this ternary? 

Could you also explain how this ternary works? I have trouble understanding it sorry.

Reply:
With this conditional type, the return type will always be of the form of the attributes you asked. Typescript can't know what is occurring at runtime.

In my opinion, the idea of using Typescript is also to make sure that when you use `addDocuments` the type is enforced and all the properties are provided correctly. If some properties are optional like maybe `description` in your example the user should have the appropriate type:
```ts
type Movie = {
  id: string,
  title: string,
  description?: string
}
const index = client.getIndex<Movie>()
index.search('batman', {
 attributesToRetrieve: ["description", "id", "title"]
}
// => of type
{
  id:string,
  title: string,
  description: string | undefined
}
```
Will be present no matter what if the field is not optional
<img width="526" alt="Screen Shot 2020-07-15 at 11 21 37 AM" src="https://user-images.githubusercontent.com/4748419/87563938-cd94f080-c68d-11ea-8975-5e9238a40469.png">

Will be optional if the field is optional in the type
<img width="562" alt="Screen Shot 2020-07-15 at 11 24 27 AM" src="https://user-images.githubusercontent.com/4748419/87563941-ce2d8700-c68d-11ea-896e-dd163c76cebc.png">

Please see a working example [here](https://www.staging-typescript.org/play?#code/C4TwDgpgBAEglsAPAFQHxQLxWVAZFAbygH0AzAewCcBbAQ2GAgBMB+ALmygF8AoHuAHaNKpWgGNoAZQi1KYgBYAFWbWoBnFOgI8ouqPWCU4AIwCujNcnIAlCIbgQAbhHZQAogA9D4pAGsIIOSk2AA0UGr2AgDmqADaALpQAD7uXpQ+iP6BwchhEUbRqDy8-EIQIuJSMnLytmpg5AJqEChhilAQXhACTGpQ0rIKyunqmlo6UPIIahyKsQDkBkZmFla29k4Q84mdjD19WUHYE7osUACClOkgiCd6uvBId-e6inBivq2pYgA2pkwtQ45MKeX7-QEBI65KBzRYMZbmCCWGx2IybbaoTHPPRFF5QXF6WYLJYmRHI9Zo5zbDpdfYXK60G6CUjlKAAaQJp3p10Qj0Qbw+X1BfwBmUhwO+Ioh2VC7Mx8ruHEuPL5aCKJVAkCgAFlyI4HJhCHc4Ew2AJTNRjOUQndgAgfhA2PlBFE7gC1GIjGA7Y0OM7osU+GJGhEoGAVOoOAMasNVBpdfqIOgsNp7iSVki1qiHM4OAsTfMwvN3Z64N64I1C1B5nbgA7toGeMGmsAoJQkQ0mo7+tUFHVO81EAmHGFNRAjuGRmpk0b7lNgDNYqm8VAS16fQI2PNC9jdLWHVudyuoCa2ABGG3HshUOgMZhsZfH3Rrssbw+Xp97+2O7cfz+ns9d24O4uHiRsxx1PUHAAdQQeQAHly0aWgfgAMQcH4mENR8T1Nc1LWtW1vydSJXXuF8kIEVx-VdEpm1DSc4wAJijXslAjeMoIgWDgAQyiUPQiBMJnHD0zJLMNlzKB8yYKtiyRUtKLk-ctjAuiQ1bdt6hDCAWJ7QZag7HShy4ni+I3ASMKYUdwHHYJGPUJiRLuedFxw8iFPXCtN1-ICVPfICAL-F5rxoAx73clcKLfXzPy-OsfyPOKAKA3h7lA4ogA)

For the ternary no trouble, those are never easy to read unfortunately :/ I was thinking about creating a helper out of that so that it doesn't clutter the `SearchResponse` type. Could be done later.

```ts
// The second generic P is used to capture the SearchParams type
interface SearchResponse<T, P extends SearchParams<T>> { 
// P represents the SearchParams
// and by using the indexer P['attributesToRetrieve'], we're able to pick the type of `attributesToRetrieve`
// and check whether the attribute is a single key present in the generic
  hits: P['attributesToRetrieve'] extends keyof T // `attributesToRetrieve` contains one single key
    ? Array< // So we return an array of
        Hit< // hits
           // We exclude the `attributesToRetrieve` first from the generic, and then we exclude what has been returned to make sure we only Pick the `attributesToRetrieve` on the generic T
          Pick<T, Exclude<keyof T, Exclude<keyof T, P['attributesToRetrieve']>>>
        >
      >
    : P['attributesToRetrieve'] extends Array<infer K> // Otherwise if P['c'] is an array, we use `infer K` to extract the keys in the array in place
    ? Array<Hit<Pick<T, Exclude<keyof T, Exclude<keyof T, K>>>>> // Same extraction method as above when we have a single `attributesToRetrieve`
    : Array<Hit<T>> // Finally return the full type as `attributesToRetrieve` is neither a single key nor an array of keys
}

```

@bidoubiwa Let me know if this explanation is enough :) You can play with it by yourself [here](https://www.staging-typescript.org/play?#code/C4TwDgpgBAEglsAPAFQHxQLxWVAZFAbygH0AzAewCcBbAQ2GAgBMB+ALmygF8AoHuAHaNKpWgGNoAZQi1KYgBYAFWbWoBnFOgI8ouqPWCU4AIwCujNcnIAlCIbgQAbhHZQAogA9D4pAGsIIOSk2AA0UGr2AgDmqADaALpQAD7uXpQ+iP6BwchhEUbRqDy8-EIQIuJSMnLytmpg5AJqEChhilAQXhACTGpQ0rIKyunqmlo6UPIIahyKsQDkBkZmFla29k4Q84mdjD19WUHYE7osUACClOkgiCd6uvBId-e6inBivq2pYgA2pkwtQ45MKeX7-QEBI65KBzRYMZbmCCWGx2IybbaoTHPPRFF5QXF6WYLJYmRHI9Zo5zbDpdfYXK60G6CUjlKAAaQJp3p10Qj0Qbw+X1BfwBmUhwO+Ioh2VC7Mx8ruHEuPL5aCKJVAkCgAFlyI4HJhCHc4Ew2AJTNRjOUQndgAgfhA2PlBFE7gC1GIjGA7Y0OM7osU+GJGhEoGAVOoOAMasNVBpdfqIOgsNp7iSVki1qiHM4OAsTfMwvN3Z64N64I1C1B5nbgA7toGeMGmsAoJQkQ0mo7+tUFHVO81EAmHGFNRAjuGRmpk0bG2OdXqHAB1BDyADy5catB+ADEHD8mIbU3oTWaLVbKDb7rWHU7Iq77iWvT6BK5-a6Ss3Q5O4wAmKO9koEbxouEArsA66bgI257hAB4zseujpmSWYbLmUD5kwVbFkipZQdhN5bPEjZfq27b1CGED-j2gy1B2lFDqB4GQS+MH7kwo7gOOwQ-uov4IXcUzADMsSIS8T5li+bDzIW2K6IR0myXiuingAjFeylkFQdAMMwbBicpUASVBikaYZ8n2o6MlmeZalydwdxcMRXBAA)